src/layout-text.ts

```diff
@@ -1888,9 +1893,16 @@ export class Paragraph {
           log?.(`    Shaping "${this.string.slice(offset, end)}" with font ${match.filename}\n`);
           log?.('    Shaper returned: ' + logGlyphs(shapedPart) + '\n');
 
+
           while (!hbClusterState.done) {
             nextGlyph(hbClusterState);
 
+            // If the current font does not support emojis, should fallback to the next match font.
+            if (attrs.isEmoji && hbClusterState.needsReshape) {
```

Comment:
Need to at least check `isLastMatch` here or it will infinite loop. One of the tests does that currently.

Reply:
I add a `isLastMatch` check, finished with a tofu glyph when this is the last font match.

It will render like this with tofu glyph on macOS.
<img width="958" alt="截屏2024-04-29 12 57 01" src="https://github.com/chearon/dropflow/assets/9712830/d4f17157-c2f0-423b-b349-179b396f3a16">